### ISSUE TITLE
Support of https backend with wildcard certificates ?

### ISSUE THREAD

**Issue by nsteinmetz** [containous/traefik#2770]
### Do you want to request a *feature* or report a *bug*?

Report a bug

### What did you do?

Run nginx https endpoint in http2 mode behind traefik.

### What did you expect to see?

Access to the nginx endpoint

### What did you see instead?

`Internal Server Error`

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      v1.5.0
Codename:     cancoillotte
Go version:   go1.9.2
Built:        2018-01-23_04:42:32PM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
[entryPoints]
    [entryPoints.http]
    address = ":80"
        [entryPoints.http.redirect]
        entryPoint = "https"
    [entryPoints.https]
    address = ":443"
        [entryPoints.https.tls]
            [[entryPoints.https.tls.certificates]]
            certFile = "//wildcard.mycorp.com.crt"
            keyFile = "/wildcard.mycorp.com.key"
```

`docker-compose.yml` file is :

### If applicable, please paste the log output in debug mode (`--debug` switch)

```yml
version: '2'
services:
  traefik:
    image: traefik:cancoillotte-alpine
    command: --api --docker
    networks:
      - mycorp
    ports:
      - "80:80"
      - "443:443"
      - "8080:8080"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ${PWD}/traefik/traefik.toml:/traefik.toml
      - ${PWD}/traefik/ssl/wildcard.mycorm.com.key:/wildcard.mycorp.com.key
      - ${PWD}/traefik/ssl/wildcard.mycorp.com.crt:/wildcard.mycorp.com.crt
  nginx:
    build: nginx
    networks:
      mycorp:
        aliases:
         - docker.mycorp.com
    labels:
      traefik.frontend.rule: "Host:docker.mycorp.com"
      traefik.frontend.entryPoints: "http,https"
      traefik.port: "443"
      traefik.protocol: "https"
networks:
  mycorp:
    driver: bridge
```
and `Dockerfile`:

```
FROM nginx
ADD ssl /etc/nginx/ssl
ADD default.conf /etc/nginx/conf.d/default.conf
```

and `default.conf`:

```
server {
    listen 80;
    server_name _ default_server;
    root /usr/share/nginx/html;
    index index.html;
 }
server {
     listen 443 ssl http2;
     server_name _ default_server;
     ssl_certificate /etc/nginx/ssl/wildcard.mycorp.com.chained.crt;
     ssl_certificate_key /etc/nginx/ssl/wildcard.mycorp.com.key;
    root /usr/share/nginx/html;
    index index.html;
}
```

If I switch back nginx https endpoint to http/1.1, it works as expected.

Output of debug mode:

```shell
traefik_1  | time="2018-01-26T14:04:45Z" level=debug msg="vulcand/oxy/roundrobin/rr: begin ServeHttp on request" Request="{"Method":"GET","URL":{"Scheme":"","Opaque":"","User":null,"Host":"","Path":"/","RawPath":"","ForceQuery":false,"RawQuery":"","Fragment":""},"Proto":"HTTP/2.0","ProtoMajor":2,"ProtoMinor":0,"Header":{"Accept":["*/*"],"User-Agent":["curl/7.54.0"]},"ContentLength":0,"TransferEncoding":null,"Host":"docker.mycorp.com","Form":null,"PostForm":null,"MultipartForm":null,"Trailer":null,"RemoteAddr":"172.18.0.1:36652","RequestURI":"/","TLS":null}"
traefik_1  | time="2018-01-26T14:04:45Z" level=debug msg="vulcand/oxy/roundrobin/rr: Forwarding this request to URL" Request="{"Method":"GET","URL":{"Scheme":"","Opaque":"","User":null,"Host":"","Path":"/","RawPath":"","ForceQuery":false,"RawQuery":"","Fragment":""},"Proto":"HTTP/2.0","ProtoMajor":2,"ProtoMinor":0,"Header":{"Accept":["*/*"],"User-Agent":["curl/7.54.0"]},"ContentLength":0,"TransferEncoding":null,"Host":"docker.mycorp.com","Form":null,"PostForm":null,"MultipartForm":null,"Trailer":null,"RemoteAddr":"172.18.0.1:36652","RequestURI":"/","TLS":null}" ForwardURL=https://172.18.0.3:443
traefik_1  | time="2018-01-26T14:04:45Z" level=debug msg="vulcand/oxy/forward: begin ServeHttp on request" Request="{"Method":"GET","URL":{"Scheme":"https","Opaque":"","User":null,"Host":"172.18.0.3:443","Path":"","RawPath":"","ForceQuery":false,"RawQuery":"","Fragment":""},"Proto":"HTTP/2.0","ProtoMajor":2,"ProtoMinor":0,"Header":{"Accept":["*/*"],"User-Agent":["curl/7.54.0"]},"ContentLength":0,"TransferEncoding":null,"Host":"docker.mycorp.com","Form":null,"PostForm":null,"MultipartForm":null,"Trailer":null,"RemoteAddr":"172.18.0.1:36652","RequestURI":"/","TLS":null}"
traefik_1  | time="2018-01-26T14:04:45Z" level=debug msg="vulcand/oxy/forward/http: begin ServeHttp on request" Request="{"Method":"GET","URL":{"Scheme":"https","Opaque":"","User":null,"Host":"172.18.0.3:443","Path":"","RawPath":"","ForceQuery":false,"RawQuery":"","Fragment":""},"Proto":"HTTP/2.0","ProtoMajor":2,"ProtoMinor":0,"Header":{"Accept":["*/*"],"User-Agent":["curl/7.54.0"]},"ContentLength":0,"TransferEncoding":null,"Host":"docker.mycorp.com","Form":null,"PostForm":null,"MultipartForm":null,"Trailer":null,"RemoteAddr":"172.18.0.1:36652","RequestURI":"/","TLS":null}"
traefik_1  | time="2018-01-26T14:04:45Z" level=info msg="vulcand/oxy/forward/http: Round trip: https://172.18.0.3:443, code: 500, Length: 21, duration: 4.925836ms tls:version: 303, tls:resume:false, tls:csuite:c02f, tls:server:docker.mycorp.com"
traefik_1  | time="2018-01-26T14:04:45Z" level=debug msg="vulcand/oxy/forward/http: completed ServeHttp on request" Request="{"Method":"GET","URL":{"Scheme":"https","Opaque":"","User":null,"Host":"172.18.0.3:443","Path":"","RawPath":"","ForceQuery":false,"RawQuery":"","Fragment":""},"Proto":"HTTP/2.0","ProtoMajor":2,"ProtoMinor":0,"Header":{"Accept":["*/*"],"User-Agent":["curl/7.54.0"]},"ContentLength":0,"TransferEncoding":null,"Host":"docker.mycorp.com","Form":null,"PostForm":null,"MultipartForm":null,"Trailer":null,"RemoteAddr":"172.18.0.1:36652","RequestURI":"/","TLS":null}"
traefik_1  | time="2018-01-26T14:04:45Z" level=debug msg="vulcand/oxy/forward: completed ServeHttp on request" Request="{"Method":"GET","URL":{"Scheme":"https","Opaque":"","User":null,"Host":"172.18.0.3:443","Path":"","RawPath":"","ForceQuery":false,"RawQuery":"","Fragment":""},"Proto":"HTTP/2.0","ProtoMajor":2,"ProtoMinor":0,"Header":{"Accept":["*/*"],"User-Agent":["curl/7.54.0"]},"ContentLength":0,"TransferEncoding":null,"Host":"docker.mycorp.com","Form":null,"PostForm":null,"MultipartForm":null,"Trailer":null,"RemoteAddr":"172.18.0.1:36652","RequestURI":"/","TLS":null}"
traefik_1  | time="2018-01-26T14:04:45Z" level=debug msg="vulcand/oxy/roundrobin/rr: competed ServeHttp on request" Request="{"Method":"GET","URL":{"Scheme":"","Opaque":"","User":null,"Host":"","Path":"/","RawPath":"","ForceQuery":false,"RawQuery":"","Fragment":""},"Proto":"HTTP/2.0","ProtoMajor":2,"ProtoMinor":0,"Header":{"Accept":["*/*"],"User-Agent":["curl/7.54.0"]},"ContentLength":0,"TransferEncoding":null,"Host":"docker.mycorp.com","Form":null,"PostForm":null,"MultipartForm":null,"Trailer":null,"RemoteAddr":"172.18.0.1:36652","RequestURI":"/","TLS":null}"
```

What is really weird is that when you connect on traefik container (alpine mode), you can `curl https://docker.mycorp.com` and it works as expected...

Let me know if you need more details.

**Comment by Juliens** [comment]
Are you sure your backend certificate is valid ?
May be you need to add your certificate in RootCAs, or use InsecureSkipVerify (https://docs.traefik.io/configuration/commons/)

**Comment by nsteinmetz** [comment]
@Juliens yes they are valid, if I run nginx image alone with certificates, it works as expected.

And also as I said, if I connect to the container and do `curl https://docker.mycorp.com`, it works!

**Comment by nsteinmetz** [comment]
@Juliens 

If I add in the toml:

```
InsecureSkipVerify = true

[entryPoints]
    [entryPoints.http]
    address = ":80"
        [entryPoints.http.redirect]
        entryPoint = "https"
    [entryPoints.https]
    address = ":443"
        [entryPoints.https.tls]
            [[entryPoints.https.tls.certificates]]
            certFile = "//wildcard.mycorp.com.crt"
            keyFile = "/wildcard.mycorp.com.key"
```

It works.

**Comment by nsteinmetz** [comment]
With `RootCAs`, what is expected, only the final cert? a combined one ? a full one ?

```
#InsecureSkipVerify = true
RootCAs = [ "/wildcard.mycorp.com.crt" ]

[entryPoints]
    [entryPoints.http]
    address = ":80"
        [entryPoints.http.redirect]
        entryPoint = "https"
    [entryPoints.https]
    address = ":443"
        [entryPoints.https.tls]
            [[entryPoints.https.tls.certificates]]
            certFile = "//wildcard.mycorp.com.crt"
            keyFile = "/wildcard.mycorp.com.key"
```

**Comment by Juliens** [comment]
May be a typo ? ( `certFile = "//wildcart.."` )

It seems to be a problem with your configuration, can you go to [slack](https://traefik.herokuapp.com/) for this.

**Comment by nsteinmetz** [comment]
@Juliens thanks for the typo. The issue is not between the browser and traefik. It works.

It's really between traefik and the https endpoint (nginx) behind traefik.

Just tried with:

```
#InsecureSkipVerify = true
RootCAs = [ "/wildcard.mycorp.com.crt", "/wildcard.mycorp.com.chained.crt" ]
```
=> KO

```
#InsecureSkipVerify = true
RootCAs = ["/wildcard.mycorp.com.chained.crt" ]
```

=> KO

So the only working configuration so far is :

```
InsecureSkipVerify = true
```

=> OK

**Comment by ldez** [comment]
Thanks for your interest in Traefik !

Every question asked in issues means less features 😉.

The issue tracker is for reporting bugs and feature requests only.
For end-user related support questions, refer to one of the following:

- [Stack Overflow](https://stackoverflow.com/questions/tagged/traefik) using the "traefik" tag
- the [Traefik community Slack channel](https://traefik.herokuapp.com)

**Comment by nsteinmetz** [comment]
@ldez : I agree, I thouht it was an http2 issue first. I still think there is an issue about support of https backend.

**Comment by ldez** [comment]
come on Slack and if we think it's a bug I will re-open 😉

**Comment by nsteinmetz** [comment]
Just did it :)

**Comment by nsteinmetz** [comment]
TL;DR; Traefik will use IP and not host - so certificates will never validate:

```
curl https://172.18.0.2/
curl: (51) SSL: no alternative certificate subject name matches target host name '172.18.0.2'
/ # curl --header "Host: docker.mycorp.com" https://172.18.0.2/
curl: (51) SSL: no alternative certificate subject name matches target host name '172.18.0.2'
/ # curl --header "Host:docker.mycorp.com" https://172.18.0.2/
curl: (51) SSL: no alternative certificate subject name matches target host name '172.18.0.2'
```

whereas:

```
curl -I https://docker.mycorp.com
HTTP/1.1 200 OK
Server: nginx/1.12.2
Date: Fri, 26 Jan 2018 15:20:21 GMT
Content-Type: text/html
Content-Length: 612
Last-Modified: Tue, 11 Jul 2017 13:29:18 GMT
Connection: keep-alive
ETag: "5964d2ae-264"
Accept-Ranges: bytes
```

So only solution either to use `InsecureSkipVerify = true` or that certificates will be update to match container IPs

**Comment by Yivan** [comment]
Hello,

I just got the same problem `Internal Server Error` if i activate `traefik.protocol=https` and `traefik.port=443` on my docker container.

Exactly same setup work great with jwidler/nginx-proxy for instance. Certificates on the container (installed with apache 2.4) are real signed one (i installed them on traefic and on the apache of my container). If i request directly my apache container with https://... all browsers say certificat is valid (green). So the certificats in the container are ok.

Finaly is there any solution for production ? Using `InsecureSkipVerify = true` is not safe.

**Comment by ldez** [comment]
@Yivan Thanks for your interest in Traefik 😃 

This issue is closed. Please discuss this in :
- [Stack Overflow](https://stackoverflow.com/questions/tagged/traefik) using the "traefik" tag
- the [Traefik community Slack](https://traefik.herokuapp.com)

**Comment by Yivan** [comment]
Thanks @ldez 
For people interrested, discussion is here:
https://stackoverflow.com/questions/49412376/internal-server-error-with-traefik-https-backend-on-port-443

**Comment by arno01** [comment]
In case if someone needs it, here is the script I eventually ended up with https://git.nixaid.com/arno/gencert
Without ``InsecureSkipVerify = true``.
